DhsTranslator/src/IDhsAdapter.h

```diff
@@ -0,0 +1,277 @@
+            if (type == DHS_DT_STRING && value.stringVal != NULL) {
+                delete value.stringVal;
+            }
+        }
+        const std::string& getName() const {
+            return name;
+        }
+        DHS_DATA_TYPE getType() const {
+            return type;
+        }
+        template<typename T> T getValue() const {
+            throw(std::logic_error(
+                        "Attempt to retrieve unsupported type of value from Keyword object."));
+        }
+    private:
+        std::string name;
+        DHS_DATA_TYPE type;
+        union {
+            bool boolVal;
+            unsigned long int uintVal;
+            long int intVal;
+            float floatVal;
+            double doubleVal;
+            std::string *stringVal;
+        } value;
+    };
+
+    /**
+     * Request an unique image ID to the DHS Server. The caller is responsible
+     * for providing the object that will receive the image ID, and to
+     * dispose of it.
+     *
+     * @param id pointer to object that will receive the image ID.
+     * @return DHS_S_SUCCESS if the operation was successful, or the error code
+     * otherwise.
+     */
+    virtual DHS_STATUS createImage(ImageId& id) = 0;
+
+    /**
+     * Configure the life time of an image.
+     *
+     * @param id pointer the image ID.
+     * @param lifeTime type of life time.
+     * @return DHS_S_SUCCESS if the operation was successful, or the error code
+     * otherwise.
+     */
+    virtual DHS_STATUS setImageLifeTime(const ImageId& id,
+            DHS_BD_LIFETIME lifeTime) = 0;
+
+    /**
+     * Set the list of contributor for an image.
+     *
+     * @param id pointer the image ID.
+     * @param contribs vector with the names of the contributors.
+     * @return DHS_S_SUCCESS if the operation was successful, or the error code
+     * otherwise.
+     */
+    virtual DHS_STATUS setImageContrib(const ImageId& id,
+            const std::vector<std::string>& contribs) = 0;
+
+    /**
+     * Set a group of keywords in an image.
+     *
+     * @param id pointer the image ID.
+     * @param keywords a vector of keywords to be set.
+     * @param final indicates if this is the last piece of the image.
+     * @return DHS_S_SUCCESS if the operation was successful, or the error code
+     * otherwise.
+     */
+    virtual DHS_STATUS setImageKeywords(const ImageId& id,
+            const std::vector<Keyword>& keywords, bool final = false) = 0;
+
+    /**
+     * Set the timeout for the blocking operations.
+     *
+     * @param timeout maximum time to wait for completion, in milliseconds.
+     */
+    virtual void setTimeout(unsigned int timeout) = 0;
+
+    virtual ~IDhsAdapter() {
+    }
+
+};
+
+template<> inline int8_t IDhsAdapter::Keyword::getValue<int8_t>() const {
+    if (type == DHS_DT_INT8) {
+        return value.intVal;
+    }
+    throw(std::logic_error(
+            "Attempt to retrieve int8_t value from Keyword of a different type."));
```

Comment:
Since you know the wrong type that was used, why don't put it in the error message? Probably can be useful if this error ever happens.